File: BinaryTree/CPP/find_a_corresponding_node_of_a_binary_tree_in_a_clone_of_that_tree.cpp
#include <stdio.h>
void *p = NULL;

// Definition for a binary tree node.
struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};


class Solution {
public:
    TreeNode* getTargetCopy(TreeNode* original, TreeNode* cloned, TreeNode* target) {
        return inorder(original, cloned, target);
    }

    TreeNode* inorder(TreeNode* o, TreeNode* c, TreeNode* t) {
        if (o == nullptr) return nullptr;
        if (o == t) return c;

        // Search in the left subtree
        TreeNode* left = inorder(o->left, c->left, t);
        if (left != nullptr) return left;

        // Search in the right subtree
        return inorder(o->right, c->right, t);
    }
};

File: BinaryTree/find_a_corresponding_node_of_a_binary_tree_in_a_clone_of_that_tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def getTargetCopy(self, original: TreeNode, cloned: TreeNode, target: TreeNode) -> TreeNode:
        if not original: return None
        if original == target: return cloned
        r = self.getTargetCopy(original.right, cloned.right, target)
        if r: return r
        return self.getTargetCopy(original.left, cloned.left, target)


    #  another implementation
    def getTargetCopy(self, original: TreeNode, cloned: TreeNode, target: TreeNode) -> TreeNode:
        def helper(o: TreeNode, c: TreeNode) -> None:
            if o:
                helper(o.left, c.left)
                if o is target: self.ans = c
                helper(o.right, c.right)
        helper(original, cloned)
        return self.ans